### PR TITLE
Save on Finish

### PR DESCRIPTION
**Pull request by JeremySkinner**
Adds a new config field to control whether the form entities should be saved at the end of each step (default) or once the whole form is complete.

### CODE REVIEW

File: contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php [hunk 1]
```diff
@@ -52,13 +54,18 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
...
 
     $form_display = $form_state->get('form_display');
     $form_display->extractFormValues($entity, $form, $form_state);
 
     $form_display->formSubmitComponents($form, $form_state);
-    $entity->save();
+    if (!$config->shouldSaveOnFinish()) {
+      $entity->save();
+    }
     $form_display->saveFormEntities($form, $form_state);
```
Comment (rlmumford) [inline contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php]: I think this is going to save any entities added to the step that aren't the 'base entity'. As way to test would be to add the Node Author to the first page form then see if any changes made to the author were saved.

That may be where we need to pass the alternative FormEntityManager.

**Review by rlmumford — CHANGES_REQUESTED** [review]

File: contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php (hunk 1 above)
Comment (JeremySkinner) [reply contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php]: Thanks, I'll investigate

**Comment by JeremySkinner** [comment]
@rlmumford It turns out this is simpler than we thought. I didn't end up using the `TempstoreAwareFlexiformFormEntityManager` in the end. 

As discussed yesterday, it made more sense to store the flexiform entities in `$cached_values`, and then retrieve them from here and pass them to the `$provided` collection on form rebuild. This means the regular `FlexiformFormEntityManager` could still be used. 

I can re-introduce the entity manager factory if you'd like, although I think it probably makes sense to keep it simple for now. 

Please take another look and let me know if you'd like anything changed. 

Sorry this took so long!

File: contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php [hunk 2]
```diff
@@ -52,14 +67,30 @@ public function validateForm(array &$form, FormStateInterface $form_state) {
...
+      $form_display->saveFormEntities($form, $form_state);
+    }
+    else {
+      // Only saving at the end - cache the other form entities for now.
+      $flexiform_entities = array_map(function ($wrapper) {
+        return $wrapper->getEntity();
+      }, $form_display->getFormEntityManager()->getFormEntities());
+
+      $cached_values = $form_state->getTemporaryValue('wizard');
+      $cached_values['flexiform_entities'][$step] = $flexiform_entities;
```
Comment (rlmumford) [inline contrib/wizard/src/Plugin/WizardStep/EntityFormMode.php:91]: @JeremySkinner what behaviour does this cause if you want to use the same entity across multiple steps?

It looks like the changes on step 1 might not be included on step 2 (or whatever step had that entity a second time). I'm trying to think of a way we could test that. Hmmm.

**Review by rlmumford — APPROVED** [review]

**Pull request merged**